AVX2 implementation of DMVR SAD for VVC

stone-d-chen opened the pull request:
Adds AVX2 assembly for SAD used in DMVR (decoder-side motion vector refinement). The main difference is that in VVC, SAD is only calculated on even rows of the PU to reduce complexity. Implements SAD via min/max/sub for 16bit values.

DMVR is restricted to PUs whose width >= 8, height >=8 and width * height >= 128 (ie 8x8 is not a valid size).

Based on https://github.com/ffvvc/FFmpeg/pull/213#issue-2246959529 but on top of ffmpeg/master

nuomi2021 commented:
@QSXW  please help review.
thank you

nuomi2021 commented on libavcodec/vvc/inter_template.c:
```diff
@@ -458,7 +458,12 @@ static void FUNC(ff_vvc_inter_dsp_init)(VVCInterDSPContext *const inter)
     inter->apply_prof_uni_w     = FUNC(apply_prof_uni_w);
     inter->apply_bdof           = FUNC(apply_bdof);
     inter->prof_grad_filter     = FUNC(prof_grad_filter);
-    inter->sad                  = vvc_sad;
```
you can use a loop for this
for (int i = 0; i < FF_ARRAY_ELEMS(inter->sad); I++)

stone-d-chen replied:
I've made the change here https://github.com/ffvvc/FFmpeg/blob/2ee95457a35b0701716ed17d27c967163c1ec3ea/libavcodec/vvc/inter_template.c#L462-L464

nuomi2021 commented:
better split to 3 commits
1. refact the current code.
2. add avx2
3. provide check asm and benchmark data.
thank you

QSXW commented on libavcodec/x86/vvc/vvcdsp_init.c:
```diff
@@ -233,19 +249,23 @@ void ff_vvc_dsp_init_x86(VVCDSPContext *const c, const int bd)
                 break;
             case 12:
                 AVG_INIT(12, avx2);
+                SAD_INIT();
```
If SAD_INIT is available for all bit depth, we can move it above the switch.

stone-d-chen replied:
I've moved this above the switch
https://github.com/stone-d-chen/ffvvc/blob/0c68dad30567785304760ef6aeb2a3e439a4d6a2/libavcodec/x86/vvc/vvcdsp_init.c#L257

QSXW commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,265 @@
+%endmacro
+
+
+INIT_YMM avx2
```
Generally, FFmpge will wrap the cglobal function by a macro and init them at the end of the file. You can refer to this 
https://github.com/FFmpeg/FFmpeg/blob/bba996d6cdac15367f265e245987477d0f7b1899/libavcodec/x86/vvc/vvc_mc.asm#L299.

stone-d-chen replied:
I've created the macro here 
https://github.com/stone-d-chen/ffvvc/blob/0c68dad30567785304760ef6aeb2a3e439a4d6a2/libavcodec/x86/vvc/vvc_sad.asm#L72

QSXW replied:
Could you update it like https://github.com/QSXW/FFmpeg-VVC/blob/cfbdc27005f067951cba8a2499af3ae19b779ae3/libavcodec/x86/vvc/vvc_alf.asm#L801?

QSXW commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,265 @@
+%endmacro
+
+%macro HORIZONTAL_ADD 3  ; xm0, xm1, m1
+    vextracti128      %1, %3 , q0001 ;        3        2      1          0
```
There is a redundant space after the %3

QSXW commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,265 @@
+
+    lea             %1, [%1 + %5 * 2]
+    lea             %2, [%2 + %6 * 2]
+
```
Better to remove these tw9 blank line.

QSXW commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,265 @@
+
+    RET
+
+cglobal vvc_sad_128_16bpc, 6, 9, 13, src1, src2, dx, dy, block_w, block_h, off1, off2, row_idx
```
looks like there are a lot of the same operations of 32/64/128? Can we merge them into one macro?

stone-d-chen replied:
Created the macro here 
https://github.com/stone-d-chen/ffvvc/blob/0c68dad30567785304760ef6aeb2a3e439a4d6a2/libavcodec/x86/vvc/vvc_sad.asm#L72

QSXW commented on tests/checkasm/vvc_sad.c:
```diff
@@ -0,0 +1,191 @@
+#include <string.h>
```
The license header is missing. Please refer to https://github.com/FFmpeg/FFmpeg/blob/bba996d6cdac15367f265e245987477d0f7b1899/tests/checkasm/vvc_mc.c#L2

QSXW commented on tests/checkasm/vvc_sad.c:
```diff
@@ -0,0 +1,191 @@
+    declare_func_emms(AV_CPU_FLAG_AVX2,int, const int16_t *src0, const int16_t *src1, int dx, int dy, int block_w, int block_h);
+
+    if(check_func(c.inter.sad[5], "vvc_sad_128_16bpc"))
+    {
```
The FFmpeg use the following style
```
if (condition) {
}
```
Could you help update them?

stone-d-chen commented:
Hi @nuomi2021 and @QSXW 

Thanks for the feedback! I've forced pushed my changes addressing the feedback.

> better split to 3 commits
> 
> refact the current code.
> add avx2
> provide check asm and benchmark data.
> thank you

I split into the 3 commits suggested and attached the benchmark data to the checkasm commit message.

@QSXW 
I've addressed all the formatting changes and commented above with the new lines where I've addressed the macro changes. I've additionally added some %define's like vvc_mc.asm to clean to improve the clarity of the "magic" numbers.

I've done some changes vvc_sad.c to remove mixed declaration and code warnings (no warning anymore).

QSXW commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,201 @@
+.loop_height:
+    mov               off1q, src1q
+    mov               off2q, src2q
+    mov            row_idxd, %1 / 32
```
Looks like it is the only difference. We may have to use a C function wrapper to pass row_idxd as a parameter because the macro will introduce a copy of the data block for these repeat instructions.

stone-d-chen replied:
We can just calculate based off of block_w now that I think about it... and then just have 1 asm function for 32 to 128.

stone-d-chen replied:
Something like

```
shr   block_wd,  5
.loop_row
    ....
    dec block_wd
    jg   .loop_row
```

QSXW replied:
Sure. That's great!

stone-d-chen replied:
I've changed the function here https://github.com/ffvvc/FFmpeg/blob/680eec9938481971b393d92f37f8cf8776f43b6e/libavcodec/x86/vvc/vvc_sad.asm#L142

Just a single function now and removed the macros since they're unnecessary now

nuomi2021 commented on tests/checkasm/vvc_sad.c:
```diff
@@ -0,0 +1,201 @@
+    } while (0)
+
+
+static void check_vvc_sad_8_16bpc(void)
```
similar functions,  could we combine them and put entire file to vvc_mc.c

stone-d-chen replied:
I've moved the checkasms into 1 function in vvc_mc.c
https://github.com/ffvvc/FFmpeg/blob/2ee95457a35b0701716ed17d27c967163c1ec3ea/tests/checkasm/vvc_mc.c#L325

nuomi2021 commented on libavcodec/x86/vvc/vvcdsp_init.c:
```diff
@@ -204,6 +204,18 @@ AVG_FUNCS(16, 12, avx2)
+int ff_vvc_sad_32_128_16bpc_avx2(const int16_t *src0, const int16_t *src1, int dx, int dy, int block_w, int block_h);
+
+#define SAD_INIT() do {                                                 \
+    c->inter.sad[1] = ff_vvc_sad_8_16bpc_avx2;                          \
```
Maybe we do not need 16bpc suffix. because we always handle 16bits data.

stone-d-chen replied:
I've removed "_16bpc"

nuomi2021 commented on libavcodec/x86/vvc/vvcdsp_init.c:
```diff
@@ -217,6 +229,7 @@ void ff_vvc_dsp_init_x86(VVCDSPContext *const c, const int bd)
         }
         if (EXTERNAL_AVX2_FAST(cpu_flags)) {
             MC_LINKS_AVX2(8);
+            SAD_INIT();
```
I think the asm can only work for 16bpc? not 8bpc, right?

stone-d-chen replied:
Yes however iirc 8bit video that uses DMVR has the required values converted to 16bits at some point. I think I checked this in some of the v1 conformance videos but I can double check.

stone-d-chen replied:
Example

```
conformance/passed/v1/8b420_A_Bytedance_2.bit

Samples: 14K of event 'cycles:Pu', Event count (approx.): 4030503793
  Children      Self  Command       Shared O  Symbol
+    0.80%     0.80%  vf#0:0        ffmpeg_g  [.] ff_sad16_sse2
     0.62%     0.62%  ffmpeg_g      ffmpeg_g  [.] ff_vvc_sad_16_16bpc_avx2
     0.22%     0.22%  vf#0:0        ffmpeg_g  [.] ff_sad8_mmxext
     0.21%     0.21%  vf#0:0        ffmpeg_g  [.] sad_hpel_motion_search
```

nuomi2021 commented on libavcodec/x86/vvc/vvcdsp_init.c:
```diff
@@ -233,10 +247,12 @@ void ff_vvc_dsp_init_x86(VVCDSPContext *const c, const int bd)
     }
 
     if (EXTERNAL_AVX2(cpu_flags)) {
+        SAD_INIT();
```
this not needed since you init them in bd 8, 10, 12.

stone-d-chen replied:
Sounds good, I guess one thing I'm not sure about is what the difference between EXTERNAL_AVX2 and EXTERNAL_AVX2_FAST ?

nuomi2021 replied:
Not clear, maybe some machines are not fast.
If we do not have 8, 10, 12 bit versions of sad, maybe we can init it only once

nuomi2021 reviewed (commented):
sorry, I forget submit my review comments

nuomi2021 commented:
@stone-d-chen , please also cherry pick https://github.com/ffvvc/FFmpeg/tree/workflow to make sure all conformance test passed like https://github.com/ffvvc/FFmpeg/actions/runs/8766181658/job/24058026459?pr=216

stone-d-chen committed:
.gitworkflow

Cherry-pick git workflow

stone-d-chen commented:
>  please also cherry pick https://github.com/ffvvc/FFmpeg/tree/workflow to make sure all conformance test passed like https://github.com/ffvvc/FFmpeg/actions/runs/8766181658/job/24058026459?pr=216

I've cherry-picked and pushed the commit

nuomi2021 commented:
> Before:
> BQTerrace_1920x1080_60_10_420_22_RA.vvc | 80.7 |
> Chimera_8bit_1080P_1000_frames.vvc | 158.0 |
> NovosobornayaSquare_1920x1080.bin | 159.7 |
> RitualDance_1920x1080_60_10_420_37_RA.266 | 146.3 |
> 
> After:
> BQTerrace_1920x1080_60_10_420_22_RA.vvc | 82.7 |
> Chimera_8bit_1080P_1000_frames.vvc | 167.0 |
> NovosobornayaSquare_1920x1080.bin | 166.3 |
> RitualDance_1920x1080_60_10_420_37_RA.266 | 154.0 |

👍, please move this to the second commit(the asm code one)

nuomi2021 commented on libavcodec/x86/vvc/vvcdsp_init.c:
```diff
@@ -204,6 +204,18 @@ AVG_FUNCS(16, 12, avx2)
+int ff_vvc_sad_32_128_avx2(const int16_t *src0, const int16_t *src1, int dx, int dy, int block_w, int block_h);
+
+#define SAD_INIT() do {                                           \
+    c->inter.sad[1] = ff_vvc_sad_8_avx2;                          \
```
why ff_vvc_sad_4_avx2 is not needed?

stone-d-chen replied:
For DMVR it's only valid for widths >= 8, height >= 8

stone-d-chen replied:
![image](https://github.com/ffvvc/FFmpeg/assets/15571478/048d873f-acca-49c3-b7ec-0221581a1e10)
https://ieeexplore.ieee.org/document/9252910/

nuomi2021 commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,195 @@
+    mov            row_idxd, block_wd
+    sar            row_idxd, 5
+
+    .loop_row:
```
Maybe name it as loop_width?

nuomi2021 commented on tests/checkasm/vvc_mc.c:
```diff
@@ -322,8 +322,46 @@ static void check_avg(void)
+        }
+     }
+
+    report("check_vvc_sad_8");
```
this may need to change to sad

stone-d-chen commented:
Hi @nuomi2021, 

I've pushed the suggested change.

I've also rebased on the latest ffvvc/up and tested conformance on the latest tests and the failing test passes now locally.

```
    dpb_max_num_reorder_pics.266
    sintel_120.266
    tiles_720p5994_stockholm_ter.266

total = 315, passed = 315, failed = 0, skipped = 0
```

QSXW commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -0,0 +1,195 @@
+        vpaddd              m3, m12
+
+        add             src1q, MAX_PB_SIZE * 4 * ROWS * 2 
+        add             src2q, MAX_PB_SIZE * 4 * ROWS * 2
```
align all of the `,` will be better.

stone-d-chen replied:
Re: the macros, I ended up not needing them since we could do it in one function.

The alignment should be fixed now

QSXW replied:
Okay. That's fine.

stone-d-chen commented:
Hi @QSXW 

I received a question on the ML about why not switching on block width inside the asm. I pushed a recent change that adds vvc_sad that merges together vvc_sad_8, vvc_sad_16, vvc_sad_32_128 and jumps based on the block width.

Let me know what you think and I will clean up this branch and then resubmit to the ML.

nuomi2021 commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -189,5 +189,125 @@ cglobal vvc_sad_32_128, 6, 9, 13, src1, src2, dx, dy, block_w, block_h, off1, of
+        sar            row_idxd, 5
+
+        .loop_width:
+            movu              xm0, [src1q]
```
could we load 32 bytes a time. 
so we only need to implement vvc_sad_16_128

nuomi2021 commented on libavcodec/x86/vvc/vvcdsp_init.c:
```diff
@@ -253,16 +253,14 @@ AVG_FUNCS(16, 12, avx2)
+int ff_vvc_sad_avx2(const int16_t *src0, const int16_t *src1, int dx, int dy, int block_w, int block_h);
+
+#define SAD_INIT() do {                                    \
+    c->inter.sad[1] = ff_vvc_sad_avx2;                     \
```
a loop may be better

nuomi2021 replied:
or you do not need change to sad[], just use sad since you only have one fuction

stone-d-chen replied:
yeah that's the plan is to just remove it completely like what andreas mentioned on the ML

nuomi2021 commented on libavcodec/x86/vvc/vvc_sad.asm:
```diff
@@ -189,5 +189,125 @@ cglobal vvc_sad_32_128, 6, 9, 13, src1, src2, dx, dy, block_w, block_h, off1, of
+
+    vvc_sad_8:
+        .loop_height:
+        movu              xm0, [src1q]
```
how about we process two lines at a time? if needed handle the last line at the end.
you can reuse a function(macro) for width 8 and 16.  
%macro VVC_SAD
loop_height:
    LOAD
   SAD
   DEC
   jmp loop_height
%end macro

QSXW commented:
> Hi @QSXW
> 
> I received a question on the ML about why not switching on block width inside the asm. I pushed a recent change that adds vvc_sad that merges together vvc_sad_8, vvc_sad_16, vvc_sad_32_128 and jumps based on the block width.
> 
> Let me know what you think and I will clean up this branch and then resubmit to the ML.

Sure. That's what dav1d did. It can be changed into one function and jump based on the block width.

stone-d-chen commented:
I've pushed the new version with a single function and merging 16_128 into one code path within. I'll send the updated version to the ML soon-ish

QSXW commented:
Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?

stone-d-chen commented:
> Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?

The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.

https://www.nasm.us/doc/nasmdoc3.html#section-3.9

QSXW commented:
> > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> 
> The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> 
> https://www.nasm.us/doc/nasmdoc3.html#section-3.9

Great. Please send the v2 patch to the maillist.

stone-d-chen commented:
> > > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> > 
> > 
> > The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> > https://www.nasm.us/doc/nasmdoc3.html#section-3.9
> 
> Great. Please send the v2 patch to the maillist.

Sounds good, just sent.

QSXW commented:
> > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> 
> The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> 
> https://www.nasm.us/doc/nasmdoc3.html#section-3.9

I just realized that if we define a function in C like `void vvc_sad_8`, will it conflict with our label? Can you help confirm that?

stone-d-chen commented:
> > > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> > 
> > 
> > The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> > https://www.nasm.us/doc/nasmdoc3.html#section-3.9
> 
> I just realized that if we define a function in C like `void vvc_sad_8`, will it conflict with our label? Can you help confirm that?

Seems okay, the labels seems to be stripped from the symbol table and only the function entry point is defined

```
vvc_sad.o:     file format elf64-x86-64

SYMBOL TABLE:
0000000000000000 l    df *ABS*  0000000000000000 libavcodec/x86/vvc/vvc_sad.asm
0000000000000000 l    d  .note.GNU-stack        0000000000000000 .note.GNU-stack
0000000000000000 l    d  .note.gnu.property     0000000000000000 .note.gnu.property
0000000000000000 l    d  .text  0000000000000000 .text
0000000000000000 g     F .text  0000000000000000 .hidden ff_vvc_sad_avx2
```

QSXW commented:
> > > > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> > > 
> > > 
> > > The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> > > https://www.nasm.us/doc/nasmdoc3.html#section-3.9
> > 
> > 
> > I just realized that if we define a function in C like `void vvc_sad_8`, will it conflict with our label? Can you help confirm that?
> 
> Seems okay, the labels seems to be stripped from the symbol table and only the function entry point is defined
> 
> ```
> vvc_sad.o:     file format elf64-x86-64
> 
> SYMBOL TABLE:
> 0000000000000000 l    df *ABS*  0000000000000000 libavcodec/x86/vvc/vvc_sad.asm
> 0000000000000000 l    d  .note.GNU-stack        0000000000000000 .note.GNU-stack
> 0000000000000000 l    d  .note.gnu.property     0000000000000000 .note.gnu.property
> 0000000000000000 l    d  .text  0000000000000000 .text
> 0000000000000000 g     F .text  0000000000000000 .hidden ff_vvc_sad_avx2
> ```

Okay. Seemed that the compiler would help us strip them. How about the `./configure --disable-stripping --disable-optimizations --enable-debug`?

stone-d-chen commented:
> > > > > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> > > > 
> > > > 
> > > > The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> > > > https://www.nasm.us/doc/nasmdoc3.html#section-3.9
> > > 
> > > 
> > > I just realized that if we define a function in C like `void vvc_sad_8`, will it conflict with our label? Can you help confirm that?
> > 
> > 
> > Seems okay, the labels seems to be stripped from the symbol table and only the function entry point is defined
> > ```
> > vvc_sad.o:     file format elf64-x86-64
> > 
> > SYMBOL TABLE:
> > 0000000000000000 l    df *ABS*  0000000000000000 libavcodec/x86/vvc/vvc_sad.asm
> > 0000000000000000 l    d  .note.GNU-stack        0000000000000000 .note.GNU-stack
> > 0000000000000000 l    d  .note.gnu.property     0000000000000000 .note.gnu.property
> > 0000000000000000 l    d  .text  0000000000000000 .text
> > 0000000000000000 g     F .text  0000000000000000 .hidden ff_vvc_sad_avx2
> > ```
> 
> Okay. Seemed that the compiler would help us strip them. How about the `./configure --disable-stripping --disable-optimizations --enable-debug`?

Same objdump

```
stone@fedora:~/repos/ffvvc_stone$ objdump --syms libavcodec/x86/vvc/vvc_sad.o

libavcodec/x86/vvc/vvc_sad.o:     file format elf64-x86-64

SYMBOL TABLE:
0000000000000000 l    df *ABS*  0000000000000000 libavcodec/x86/vvc/vvc_sad.asm
0000000000000000 l    d  .note.GNU-stack        0000000000000000 .note.GNU-stack
0000000000000000 l    d  .note.gnu.property     0000000000000000 .note.gnu.property
0000000000000000 l    d  .text  0000000000000000 .text
0000000000000000 g     F .text  0000000000000000 .hidden ff_vvc_sad_avx2
```

QSXW commented:
> > > > > > Should we also add `.` to the label vvc_sad_8 and vvc_sad_16 like .loop_width?
> > > > > 
> > > > > 
> > > > > The . is syntax for a local label so it feels cleaner to have vvc_sad_8, etc not have the . in front, they will be global labels and .loop_width is the local label within.
> > > > > https://www.nasm.us/doc/nasmdoc3.html#section-3.9
> > > > 
> > > > 
> > > > I just realized that if we define a function in C like `void vvc_sad_8`, will it conflict with our label? Can you help confirm that?
> > > 
> > > 
> > > Seems okay, the labels seems to be stripped from the symbol table and only the function entry point is defined
> > > ```
> > > vvc_sad.o:     file format elf64-x86-64
> > > 
> > > SYMBOL TABLE:
> > > 0000000000000000 l    df *ABS*  0000000000000000 libavcodec/x86/vvc/vvc_sad.asm
> > > 0000000000000000 l    d  .note.GNU-stack        0000000000000000 .note.GNU-stack
> > > 0000000000000000 l    d  .note.gnu.property     0000000000000000 .note.gnu.property
> > > 0000000000000000 l    d  .text  0000000000000000 .text
> > > 0000000000000000 g     F .text  0000000000000000 .hidden ff_vvc_sad_avx2
> > > ```
> > 
> > 
> > Okay. Seemed that the compiler would help us strip them. How about the `./configure --disable-stripping --disable-optimizations --enable-debug`?
> 
> Same objdump
> 
> ```
> stone@fedora:~/repos/ffvvc_stone$ objdump --syms libavcodec/x86/vvc/vvc_sad.o
> 
> libavcodec/x86/vvc/vvc_sad.o:     file format elf64-x86-64
> 
> SYMBOL TABLE:
> 0000000000000000 l    df *ABS*  0000000000000000 libavcodec/x86/vvc/vvc_sad.asm
> 0000000000000000 l    d  .note.GNU-stack        0000000000000000 .note.GNU-stack
> 0000000000000000 l    d  .note.gnu.property     0000000000000000 .note.gnu.property
> 0000000000000000 l    d  .text  0000000000000000 .text
> 0000000000000000 g     F .text  0000000000000000 .hidden ff_vvc_sad_avx2
> ```

Great. That should be no issue here. Thanks for the confirmation.

stone-d-chen commented:
hmm saw the checkasm segfaults on windows .. will need to investigate

QSXW commented:
> hmm saw the checkasm segfaults on windows .. will need to investigate

You can use the VS to debug the checkasm. If the function crashes, it will navigate you to the assembler to check which instruction triggers the exception.

stone-d-chen commented:
> > hmm saw the checkasm segfaults on windows .. will need to investigate
> 
> You can use the VS to debug the checkasm. If the function crashes, it will navigate you to the assembler to check which instruction triggers the exception.

So the segfault only seems to occur when doing "call_new" calling the function manually doesn't result in a segfault... someone on IRC suggested 

> it could be something like using a 64bit register where the function prototype only has int32. calling manually may zero-extend or sign-extend the input, whereas call_new() intentionally puts garbage in any registers that aren't guaranteed to be meaningful.

so I'll investigate along these lines

stone-d-chen commented:
> > > hmm saw the checkasm segfaults on windows .. will need to investigate
> > 
> > 
> > You can use the VS to debug the checkasm. If the function crashes, it will navigate you to the assembler to check which instruction triggers the exception.
> 
> So the segfault only seems to occur when doing "call_new" calling the function manually doesn't result in a segfault... someone on IRC suggested
> 
> > it could be something like using a 64bit register where the function prototype only has int32. calling manually may zero-extend or sign-extend the input, whereas call_new() intentionally puts garbage in any registers that aren't guaranteed to be meaningful.
> 
> so I'll investigate along these lines

Turns out this is the case, it seems like the offset calculation requires a movsxd since the dx/dy are ints but I'm treating them as quads to do the pointer calculation

stone-d-chen commented:
Pushed the changes, confirmed working on windows, will send to ML soon.

nuomi2021 commented:
@stone-d-chen , we have merged the alf asm. remember rebase before you send to upstream.

thank you

stone-d-chen commented:
> @stone-d-chen , we have merged the alf asm. remember rebase before you send to upstream.
> 
> thank you

Hi @nuomi2021, yeah the current version here and on ML is rebased against the latest ffvvc:up so it contains the alf asm.

stone-d-chen commented:
Hi all,

Pushed another version with Ronald's suggested change here and sent to the ML.
I might have messed up the send-email, let me know if I need to re-send it.

Thanks,
Stone

QSXW commented:
> Hi all,
> 
> Pushed another version with Ronald's suggested change here and sent to the ML. I might have messed up the send-email, let me know if I need to re-send it.
> 
> Thanks, Stone

No worries. It's very common to resend patches again and again for different comments.

nuomi2021 commented:
> Hi all,
> 
> Pushed another version with Ronald's suggested change here and sent to the ML. I might have messed up the send-email, let me know if I need to re-send it.
> 
> Thanks, Stone

No worries. This is one reason why open source code quality is often better than closed source.

stone-d-chen committed:
libavcodec/x86/vvc: Add AVX2 DMVR SAD functions for VVC

Implements AVX2 DMVR (decoder-side motion vector refinement) SAD functions. DMVR SAD is only calculated if w >= 8, h >= 8, and w * h > 128. To reduce complexity, SAD is only calculated on even rows. This is calculated for all video bitdepths, but the values passed to the function are always 16bit (even if the original video bitdepth is 8). The AVX2 implementation uses min/max/sub.

Additionally this changes parameters dx and dy from int to intptr_t. This allows dx & dy to be used as pointer offsets without needing to use movsxd.

Benchmarks ( AMD 7940HS )
Before:
BQTerrace_1920x1080_60_10_420_22_RA.vvc | 106.0 |
Chimera_8bit_1080P_1000_frames.vvc | 204.3 |
NovosobornayaSquare_1920x1080.bin | 197.3 |
RitualDance_1920x1080_60_10_420_37_RA.266 | 174.0 |

After:
BQTerrace_1920x1080_60_10_420_22_RA.vvc | 109.3 |
Chimera_8bit_1080P_1000_frames.vvc | 216.0 |
NovosobornayaSquare_1920x1080.bin | 204.0|
RitualDance_1920x1080_60_10_420_37_RA.266 | 181.7 |

stone-d-chen committed:
tests/checkasm: Add check_vvc_sad to vvc_mc.c

Adds checkasm for DMVR SAD AVX2 implementation.

Benchmarks ( AMD 7940HS )
vvc_sad_8x8_c: 50.3
vvc_sad_8x8_avx2: 0.3
vvc_sad_16x16_c: 250.3
vvc_sad_16x16_avx2: 10.3
vvc_sad_32x32_c: 1020.3
vvc_sad_32x32_avx2: 60.3
vvc_sad_64x64_c: 3850.3
vvc_sad_64x64_avx2: 220.3
vvc_sad_128x128_c: 14100.3
vvc_sad_128x128_avx2: 840.3

stone-d-chen commented:
Latest version (v5) with Ronald's additional feedback pushed and sent to ML. One thing of note was changing the function signature to have intptr_t dx, dy which removed the need for movsxd. I confirmed this passes checkasm on windows as well.

nuomi2021 commented:
@stone-d-chen , don't get frustrated. Every new version helps to polish your skills.

stone-d-chen commented:
Oh yeah, I don't mind :) definitely learning a lot with each revision! Will send a new version with latest feedback soon-ish

stone-d-chen commented:
https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=11863
